pyquil/noise.py

```diff
@@ -0,0 +1,551 @@
+    """
+    Get all gate applications appearing in prog.
+
+    :param Program prog: The program
+    :return: A list of all Gates in prog (without duplicates).
+    :rtype: List[Gate]
+    """
+    return sorted({i for i in prog if isinstance(i, Gate)}, key=lambda g: g.out())
+
+
+def _get_noisy_names(gates):
+    """
+    Generate new gate names for noisy gates.
+
+    :param Sequence[Gate] gates: A list of Gate objects.
+    :return: A dictionary with keys given by the input ``gates`` and values given by new gate name
+        strings.
+    :rtype: Dict[Gate,str]
+    """
+    ret = {}
+    for g in gates:
+        key = g.name, tuple(g.params)
+        if g.name in NO_NOISE:
+            ret[g] = g.name
+        elif key in SINGLE_Q:
+            ret[g] = SINGLE_Q[key][1]
+        elif key in TWO_Q:
+            ret[g] = TWO_Q[key][1]
+        else:
+            raise ValueError("Noise model for {} not yet supported".format(g))
+    return ret
+
+
+def decoherence_noise_model(gates, T1=30e-6, T2=30e-6, gate_time_1q=50e-9,
+                            gate_time_2q=150e-09, ro_fidelity=0.95):
+    """
+    The default noise parameters
+
+    - T1 = 30 us
+    - T2 = 30 us
+    - 1q gate time = 50 ns
+    - 2q gate time = 150 ns
+
+    are currently typical for near-term devices.
+
+    This function will define new gates and add Kraus noise to these gates. It will translate
+    the input program to use the noisy version of the gates.
+
+    :param Sequence[Gate] gates: The gates to provide the noise model for.
+    :param Union[Dict[int,float],float] T1: The T1 amplitude damping time either globally or in a
+        dictionary indexed by qubit id. By default, this is 30 us.
+    :param Union[Dict[int,float],float] T2: The T2 dephasing time either globally or in a
+        dictionary indexed by qubit id. By default, this is also 30 us.
+    :param float gate_time_1q: The duration of the one-qubit gates, namely RX(+pi/2) and RX(-pi/2).
+        By default, this is 50 ns.
+    :param float gate_time_2q: The duration of the two-qubit gates, namely CZ.
+        By default, this is 150 ns.
+    :param Union[Dict[int,float],float] ro_fidelity: The readout assignment fidelity
+        :math:`F = (p(0|0) + p(1|1))/2` either globally or in a dictionary indexed by qubit id.
+    :return: A NoiseModel with the appropriate Kraus operators defined.
+    """
+    all_qubits = set(sum(([t.index for t in g.qubits] for g in gates), []))
+    if isinstance(T1, dict):
+        all_qubits.update(T1.keys())
+    if isinstance(T2, dict):
+        all_qubits.update(T2.keys())
+    if isinstance(ro_fidelity, dict):
+        all_qubits.update(ro_fidelity.keys())
+
+    if not isinstance(T1, dict):
+        T1 = {q: T1 for q in all_qubits}
+
+    if not isinstance(T2, dict):
+        T2 = {q: T2 for q in all_qubits}
+
+    if not isinstance(ro_fidelity, dict):
+        ro_fidelity = {q: ro_fidelity for q in all_qubits}
+
+    noisy_identities_1q = {
```

Comment:
why are  these called identities?

Reply:
Because they are basically noise corrupted identity gates that I then later append to other operations in order to simulate corrupted versions of those. This is a bit crude, but unless we want to do fancy things like solving Lindblad equations in pyquil this will have to do.